sympy/solvers/diophantine/diophantine.py

```diff
@@ -98,19 +99,34 @@ class DiophantineSolutionSet(set):
         {(t**2, t + 2)}
     """
 
-    def __init__(self, symbols_seq, parameters=None):
+    def __init__(self, symbols_seq, parameters=None, n_parameters=None):
         super().__init__()
 
         if not is_sequence(symbols_seq):
             raise ValueError("Symbols must be given as a sequence.")
 
         self.symbols = tuple(symbols_seq)
 
+        if n_parameters is not None:
+            self.n_parameters = n_parameters
+        else:
+            if parameters is None:
+                self.n_parameters = len(self.symbols)
+            else:
+                self.n_parameters = len(parameters)
+
         if parameters is None:
-            self.parameters = symbols('%s1:%i' % ('t', len(self.symbols) + 1), integer=True)
+            self.parameters = symbols('%s1:%i' % ('t', self.n_parameters + 1), integer=True)
         else:
             self.parameters = tuple(parameters)
 
+            if n_parameters is not None:
+                if len(self.parameters) < self.n_parameters:
```

Comment:
All the logic above with `n_parameters` and `self.n_parameters` is confusing. Is there a need for `n_parameters` to be anything other than len(self.parameters)

Reply:
I'll see if I can simplify this. If `parameters` is an optional argument then it's necessary to be able to control its length.

Reply:
Ok, I have removed `n_parameters` from `DiophantineSolutionSet` and made `parameters` a required argument. Subsequently, I added a `get_parameters` method to the `DiophantineEquationType` that can be overridden in the individual solvers. This greatly simplifies the logic of dealing with the interaction of `n_parameters` / `parameters`.

Reply:
That seems better